packages/core/getUserAvailability.ts

```diff
@@ -283,65 +284,198 @@ const getDatesBetween = (dateFrom: Dayjs, dateTo: Dayjs, period: "day" | "week"
+    endDate: limitDateTo.toDate(),
+  });
+
+  // run this first, as counting bookings should always run faster..
+  if (bookingLimits) {
+    performance.mark("bookingLimitsStart");
+    await getBusyTimesFromBookingLimits(
+      bookings,
+      bookingLimits,
+      dateFrom,
+      dateTo,
+      eventType.id,
+      limitManager
+    );
+    performance.mark("bookingLimitsEnd");
+    performance.measure(`checking booking limits took $1'`, "bookingLimitsStart", "bookingLimitsEnd");
+  }
+
+  // ..than adding up durations (especially for the whole year)
+  if (durationLimits) {
+    performance.mark("durationLimitsStart");
+    await getBusyTimesFromDurationLimits(
+      bookings,
+      durationLimits,
+      dateFrom,
+      dateTo,
+      duration,
+      eventType,
+      limitManager
+    );
+    performance.mark("durationLimitsEnd");
+    performance.measure(`checking duration limits took $1'`, "durationLimitsStart", "durationLimitsEnd");
+  }
+
+  performance.mark("limitsEnd");
+  performance.measure(`checking all limits took $1'`, "limitsStart", "limitsEnd");
+
+  return limitManager.busyTimes;
+};
+
 const getBusyTimesFromBookingLimits = async (
   bookings: EventBusyDetails[],
   bookingLimits: IntervalLimit,
   dateFrom: Dayjs,
   dateTo: Dayjs,
-  eventType: EventType | undefined
+  eventTypeId: number,
+  limitManager: LimitBusyTimeManager
 ) => {
-  const busyTimes: EventBusyDetails[] = [];
-
-  // Apply booking limit filter against our bookings
-  for (const [key, limit] of Object.entries(bookingLimits)) {
-    const limitKey = key as keyof IntervalLimit;
-
-    if (limitKey === "PER_YEAR") {
-      const yearlyBusyTime = await checkBookingLimit({
-        eventStartDate: dateFrom.toDate(),
-        limitingNumber: limit,
-        eventId: eventType?.id as number,
-        key: "PER_YEAR",
-        returnBusyTimes: true,
-      });
-      if (!yearlyBusyTime) break;
-      busyTimes.push({
-        start: yearlyBusyTime.start.toISOString(),
-        end: yearlyBusyTime.end.toISOString(),
-      });
-      break;
-    }
+  for (const key of descendingLimitKeys) {
+    const limit = bookingLimits?.[key];
+    if (!limit) continue;
+
+    const unit = intervalLimitKeyToUnit(key);
+    const periodStartDates = getPeriodStartDatesBetween(dateFrom, dateTo, unit);
+
+    for (const periodStart of periodStartDates) {
+      if (limitManager.isAlreadyBusy(periodStart, unit)) continue;
+
+      // special handling of yearly limits to improve performance
+      if (unit === "year") {
+        const yearlyBusyTime = await checkBookingLimit({
+          eventStartDate: periodStart.toDate(),
+          limitingNumber: limit,
+          eventId: eventTypeId,
+          key,
+          returnBusyTimes: true,
+        });
+        if (yearlyBusyTime) {
+          limitManager.addBusyTime(periodStart, unit);
+          if (periodStartDates.every((start) => limitManager.isAlreadyBusy(start, unit))) {
+            return;
+          }
+        }
```

Comment:
Previously, if any year was marked as busy, we would skip all other limits, even if there was an additional year to check.